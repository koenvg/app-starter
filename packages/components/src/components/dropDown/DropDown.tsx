import { DropDownItemProps, DropDownItem } from "./DropDownItem";
import { Caret } from "./Caret";
import { Button } from "../button";
import { Popper, Grow, Paper, ClickAwayListener, MenuList, ButtonProps } from "@material-ui/core";
import React, { SFC } from "react";

export interface Props {
  dropDownItems: DropDownItemProps[];
  onClick: (item: DropDownItemProps) => void;
  buttonProps?: ButtonProps;
}

export const DropDown: SFC<Props> = ({ dropDownItems, children, buttonProps, onClick }) => {
  const [anchorEl, setAnchorEl] = React.useState(null);
  const handleClick = React.useCallback(
    event => {
      if (anchorEl && anchorEl.contains(event.target)) {
        setAnchorEl(null);
      } else {
        setAnchorEl(event.currentTarget);
      }
    },
    [anchorEl],
  );

  const handleItemClick = React.useCallback(
    (param: DropDownItemProps) => {
      setAnchorEl(null);
      onClick(param);
    },
    [onClick],
  );

  const handleCloseAway = React.useCallback(
    (event: React.MouseEvent<Document, MouseEvent>) => {
      if (anchorEl.contains(event.target)) {
        return;
      }
      setAnchorEl(null);
    },
    [anchorEl],
  );

  const isActive = Boolean(anchorEl);
  return (
    <div>
      <Button
        {...buttonProps}
        aria-label="Notifications"
        aria-owns={anchorEl ? "menu-list" : null}
        aria-haspopup="true"
        onClick={handleClick}
      >
        {children}
        <Caret />
      </Button>

      <Popper open={isActive} anchorEl={anchorEl} transition disablePortal placement={"bottom-start"}>
        {() => (
          <Grow in={isActive} style={{ transformOrigin: "0 0 0" }}>
            <Paper>
              <ClickAwayListener onClickAway={handleCloseAway}>
                <MenuList role="menu">
                  {dropDownItems.map(({ id, label }) => {
                    return <DropDownItem label={label} id={id} key={id} onClick={handleItemClick} />;
                  })}
                </MenuList>
              </ClickAwayListener>
            </Paper>
          </Grow>
        )}
      </Popper>
    </div>
  );
};

DropDown.displayName = "DropDown";
