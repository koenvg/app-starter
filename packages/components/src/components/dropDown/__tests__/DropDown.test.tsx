import { DropDown, Props } from "../DropDown";
import React from "react";
import { render, fireEvent } from "@testing-library/react";

describe("DropDown", () => {
  const defaultProps: Props = {
    dropDownItems: [{ id: "First", label: "First" }],
    onClick: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render the drop down", () => {
    const { container } = render(<DropDown {...defaultProps}>Click Me</DropDown>);
    expect(container).toMatchSnapshot();
  });

  it("should open and close the dropdown when you click the button", () => {
    const { container, queryByText } = render(<DropDown {...defaultProps}>Click Me</DropDown>);

    fireEvent.click(container.querySelector("button"));

    expect(container).toMatchSnapshot();

    fireEvent.click(container.querySelector("button"));

    expect(queryByText("First")).toBeNull();

    expect(container).toMatchSnapshot();
  });

  it("should close the dropdown and call onClick when you select an item from the dropdown", () => {
    const { container, queryByText } = render(<DropDown {...defaultProps}>Click Me</DropDown>);

    fireEvent.click(container.querySelector("button"));

    fireEvent.click(queryByText("First"));

    expect(queryByText("First")).toBeNull();
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
    expect(defaultProps.onClick).toHaveBeenCalledWith({ id: "First", label: "First" });
  });

  it("should close the drop down when you click away", () => {
    const { container, queryByText } = render(<DropDown {...defaultProps}>Click Me</DropDown>);

    fireEvent.click(container.querySelector("button"));

    expect(queryByText("First")).not.toBeNull();

    fireEvent.click(container);

    expect(queryByText("First")).toBeNull();
  });
});
