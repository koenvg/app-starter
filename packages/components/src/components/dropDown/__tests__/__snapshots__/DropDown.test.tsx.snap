// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown should open and close the dropdown when you click the button 1`] = `
.c0 {
  -webkit-transition: all 150ms ease-in;
  transition: all 150ms ease-in;
  display: inline-block;
  width: 0;
  height: 0;
  margin-left: 4px;
  vertical-align: middle;
  border-top: 4px solid;
  border-right: 4px solid transparent;
  border-left: 4px solid transparent;
}

<div>
  <div>
    <button
      aria-haspopup="true"
      aria-label="Notifications"
      aria-owns="menu-list"
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Click Me
        <b
          class="c0"
        />
      </span>
    </button>
    <div
      role="tooltip"
      style="position: absolute;"
      x-placement="bottom-start"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        style="opacity: 1; transform: scale(1, 1); transform-origin: 0 0 0; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <ul
          class="MuiList-root MuiList-padding"
          role="menu"
          tabindex="-1"
        >
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
          >
            First
            <span
              class="MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDown should open and close the dropdown when you click the button 2`] = `
<div>
  <div>
    <button
      aria-haspopup="true"
      aria-label="Notifications"
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Click Me
        <b
          class="sc-AykKC eHrjBS"
        />
      </span>
    </button>
  </div>
</div>
`;

exports[`DropDown should render the drop down 1`] = `
.c0 {
  -webkit-transition: all 150ms ease-in;
  transition: all 150ms ease-in;
  display: inline-block;
  width: 0;
  height: 0;
  margin-left: 4px;
  vertical-align: middle;
  border-top: 4px solid;
  border-right: 4px solid transparent;
  border-left: 4px solid transparent;
}

<div>
  <div>
    <button
      aria-haspopup="true"
      aria-label="Notifications"
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Click Me
        <b
          class="c0"
        />
      </span>
    </button>
  </div>
</div>
`;
